file/file_prefetch_buffer.cc

```diff
@@ -248,90 +241,74 @@ void FilePrefetchBuffer::AbortAllIOs() {
     assert(s.ok());
   }
 
-  // Release io_handles.
-  if (bufs_[curr_].io_handle_ != nullptr && bufs_[curr_].del_fn_ != nullptr) {
-    DestroyAndClearIOHandle(curr_);
-  } else {
-    bufs_[curr_].async_read_in_progress_ = false;
-  }
-
-  if (bufs_[second].io_handle_ != nullptr && bufs_[second].del_fn_ != nullptr) {
-    DestroyAndClearIOHandle(second);
-  } else {
-    bufs_[second].async_read_in_progress_ = false;
+  for (auto& buf : bufs_) {
+    if (buf->io_handle_ != nullptr && buf->del_fn_ != nullptr) {
+      DestroyAndClearIOHandle(buf);
+    }
+    buf->async_read_in_progress_ = false;
   }
+  FreeEmptyBuffers();
 }
 
 // Clear the buffers if it contains outdated data. Outdated data can be
 // because previous sequential reads were read from the cache instead of these
 // buffer.
-void FilePrefetchBuffer::UpdateBuffersIfNeeded(uint64_t offset, size_t length) {
-  uint32_t second = curr_ ^ 1;
-
-  if (IsBufferOutdated(offset, curr_)) {
-    bufs_[curr_].ClearBuffer();
+void FilePrefetchBuffer::ClearOutdatedData(uint64_t offset, size_t length) {
+  while (!IsBufferQueueEmpty()) {
+    BufferInfo* buf = GetFirstBuffer();
+    // Offset is greater than this buffer's end offset.
+    if (buf->IsBufferOutdated(offset)) {
+      FreeFrontBuffer();
+    } else {
+      break;
+    }
   }
-  if (IsBufferOutdated(offset, second)) {
-    bufs_[second].ClearBuffer();
+
+  if (IsBufferQueueEmpty()) {
+    return;
   }
 
-  {
-    // In case buffers do not align, reset second buffer if requested data needs
-    // to be read in second buffer.
-    if (!bufs_[second].async_read_in_progress_ &&
-        !bufs_[curr_].async_read_in_progress_) {
-      if (DoesBufferContainData(curr_)) {
-        if (bufs_[curr_].offset_ + bufs_[curr_].buffer_.CurrentSize() !=
-            bufs_[second].offset_) {
-          if (DoesBufferContainData(second) &&
-              IsOffsetInBuffer(offset, curr_) &&
-              (offset + length >
-               bufs_[curr_].offset_ + bufs_[curr_].buffer_.CurrentSize())) {
-            bufs_[second].ClearBuffer();
+  BufferInfo* buf = GetFirstBuffer();
+
+  // In case buffers do not align, reset next buffer if requested data needs
+  // to be read in that buffer.
+  if (NumBuffersAllocated() > 1) {
+    BufferInfo* next_buf = bufs_[1];
+    if (!next_buf->async_read_in_progress_ && !buf->async_read_in_progress_ &&
+        buf->DoesBufferContainData()) {
```

Comment:
Maybe add `assert(bufs_.size() <= 2);` in the places it's assumed. I think this is one such place since it doesn't look for misalignment between consecutive `bufs_` after the second one

Reply:
I have added the check in all those places `(NumBuffersAllocated > 1)` or to return if NumBuffersAllocated() == 0 or 1.

Reply:
Here I added that condition at Line 281